lib/livebook/hubs/fly.ex

```diff
@@ -0,0 +1,98 @@
+defmodule Livebook.Hubs.Fly do
+  @moduledoc false
+  defstruct [:id, :token, :name, :color, :organization]
+
+  defmodule Organization do
+    @moduledoc false
+    defstruct [:id, :slug, :name, :type]
+
+    @type t :: %__MODULE__{
+            id: String.t(),
+            slug: String.t(),
+            name: String.t(),
+            type: String.t()
+          }
+  end
+
+  @type t :: %__MODULE__{
+          id: Livebook.Utils.id(),
+          token: String.t(),
+          name: String.t(),
+          color: Livebook.Users.User.hex_color(),
+          organization: Organization.t()
+        }
+
+  alias Livebook.Utils.HTTP
+
+  def fetch_organizations(access_token) do
+    headers = [{"Authorization", "Bearer #{access_token}"}]
+
+    query = """
+    query {
+      organizations {
+        nodes {
+          id
+          slug
+          name
+          type
+        }
+      }
+    }
+    """
+
+    body = {"application/json", Jason.encode!(%{query: query})}
+
+    with {:ok, body} <- request(headers, body) do
+      organizations =
+        for node <- body["organizations"]["nodes"] do
+          %Organization{
+            id: node["id"],
+            slug: node["slug"],
+            name: node["name"],
+            type: node["type"]
+          }
+        end
+
+      {:ok, organizations}
+    end
+  end
+
+  defp graphql_endpoint do
+    Application.get_env(:livebook, :fly_io_graphql_endpoint, "https://api.fly.io/graphql")
+  end
+
+  defp request(method \\ :post, url \\ graphql_endpoint(), headers, body) do
+    HTTP.request(method, url, headers: headers, body: body)
+    |> decode_body()
+    |> handle_response()
+  end
+
+  defp decode_body({:ok, 200, _, body}), do: {:ok, Jason.decode!(body)}
+  defp decode_body({:ok, _, _, body}), do: {:error, Jason.decode!(body)}
+  defp decode_body({:error, _} = error), do: error
+
+  defp handle_response({:ok, %{"errors" => [error]}}) do
+    case error do
+      %{"extensions" => %{"code" => code}} ->
+        {:error, "request failed with code: #{code}"}
+
+      %{"message" => message} ->
+        {:error, message}
+    end
+  end
+
+  defp handle_response({:ok, %{"data" => data}}), do: {:ok, data}
+  defp handle_response(result), do: result
```

Comment:
Here is another idea for organising this code:

```elixir
defp graphql(access_token, query) do
  headers = [{"Authorization", "Bearer #{access_token}"}]
  body = {"application/json", Jason.encode!(%{query: query})}
  url = Application.get_env(:livebook, :fly_io_graphql_endpoint, "https://api.fly.io/graphql")

  case HTTP.request(:post, graphql_endpoint(), headers: headers, body: body) do
    {:ok, 200, _, body} ->
      case Jason.decode!(body) do
        %{"data" => data} ->
          {:ok, data}

        %{"extensions" => %{"code" => code}} ->
          {:error, "request failed with code: #{code}"}

        %{"message" => message} ->
          {:error, message}
      end

    {:ok, _, _, body} ->
      {:error, Jason.decode!(body)}

    {:error, _} = error ->
      error
  end
end
```

I think by combining request, decode_body, and handle_response it's a little bit easier to understand because we don't need to jump between possible returns and available clauses at each step. It also contains other graphql-specific bits.

WDYT?

Reply:
I like this one, coz in the future we will have more queries to do